bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```diff
@@ -1338,21 +1338,39 @@ private static boolean checkIsLikeRecordType(Object sourceValue, BRecordType tar
         return true;
     }
 
-    private static boolean checkFiniteTypeAssignable(Object bRefTypeValue, BType sourceType, BFiniteType lhsType) {
-        if (bRefTypeValue == null) {
+    private static boolean checkFiniteTypeAssignable(Object sourceValue, BType sourceType, BFiniteType targetType) {
+        if (sourceValue == null) {
             // we should not reach here
             return false;
         }
 
-        for (Object valueSpaceItem : lhsType.valueSpace) {
+        for (Object valueSpaceItem : targetType.valueSpace) {
             // TODO: 8/13/19 Maryam fix for conversion
-            if (getType(valueSpaceItem).getTag() == sourceType.getTag() && valueSpaceItem.equals(bRefTypeValue)) {
+            if (isFiniteTypeValue(sourceValue, sourceType, valueSpaceItem)) {
                 return true;
             }
         }
         return false;
     }
 
+    protected static boolean isFiniteTypeValue(Object sourceValue, BType sourceType, Object valueSpaceItem) {
+        BType valueSpaceItemType = getType(valueSpaceItem);
+        if (valueSpaceItemType.getTag() > TypeTags.FLOAT_TAG) {
+            return getType(valueSpaceItem).getTag() == sourceType.getTag() && valueSpaceItem.equals(sourceValue);
```

Comment:
Can't we use the 'valueSpaceItemType' from line 1357 here?
```suggestion
            return valueSpaceItemType.getTag() == sourceType.getTag() && valueSpaceItem.equals(sourceValue);
```

Reply:
+1. Will add that in a separate pr